File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int x = matrix.length;
        int y = matrix[0].length;
        
        ArrayList<Integer> result = new ArrayList<>();
        
        if (y == 1){
            for (int i = 0; i < x; i++){
                result.add(matrix[i][0]);
            }
            return result;
        }
        
        int a = 0, b = 0, roundIndex = 0;
        
        for(int i = 0; i < x * y; i++){
            result.add(matrix[a][b]);
            
            if (a == roundIndex + 1 && b == roundIndex){
                roundIndex++;
            }
            
            // in each round,
            // 1. right : a, b++
            // 2. down : a++, b
            // 3. left : a--, b
            // 4. up : a, b--
            
            // right
            if (a == roundIndex && b < y - (roundIndex + 1)){
                b++;
            }
            // left
            else if (a == x - (roundIndex + 1) && b > roundIndex){
                b--;
            }
            // up
            else if (b == roundIndex && a > (roundIndex + 1)){
                a--;
            }
            // down
            else if (b >= y - (roundIndex + 1) && a <= x - (roundIndex + 1)){
                a++;
            }
        }
        
        return result;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        output = []

        left, right = 0, len(matrix[0])
        top, bottom = 0, len(matrix)

        while left < right and top < bottom:
            # move right ->
            for i in range(left, right):
                output.append(matrix[top][i])
            top += 1

            # move down
            for i in range(top, bottom):
                output.append(matrix[i][right - 1])
            right -= 1

            if not (left < right and top < bottom):
                break

            # move left <-
            for i in range(right - 1, left - 1, -1):
                output.append(matrix[bottom - 1][i])
            bottom -= 1

            # move upwards
            for i in range(bottom - 1, top - 1, -1):
                output.append(matrix[i][left])
            left += 1
        
        return output